securedrop/tests/functional/functional_test.py

```diff
@@ -33,19 +34,50 @@ class FunctionalTest():
 
     def _unused_port(self):
         s = socket.socket()
-        s.bind(("localhost", 0))
+        s.bind(('127.0.0.1', 0))
         port = s.getsockname()[1]
         s.close()
         return port
 
+
+    def _create_tor_browser_webdriver(self, abs_log_file_path):
+        # Read the TorBrowser location from Ansible
+        with open('/opt/.tbb_path_file') as f:
+           path_to_tbb = f.readline().strip()
+        path_to_tbb = path_to_tbb + os.path.sep + "tor-browser_en-US"
+        path_to_tbb = os.path.abspath(path_to_tbb)
+
+        # Don't use Tor when reading from localhost, and turn off private browsing.
+        # We need to turn off private browsing because we won't be able to access the browser's cookies in private browsing mode.
+        # Since we use session cookies in SD anyway (in private browsing mode all cookies are set as session cookies),
+        # this should not affect session lifetime.
+        pref_dict = {
+                     'network.proxy.no_proxies_on': '127.0.0.1',
+                     'browser.privatebrowsing.autostart': False
+                    }
+        driver = TorBrowserDriver(path_to_tbb,
+                                  pref_dict=pref_dict,
+                                  tbb_logfile_path=abs_log_file_path)
+        return driver
+
+    def _create_firefox_webdriver(self, abs_log_file_path):
+        with open(abs_log_file_path, 'a') as f:
+            firefox = firefox_binary.FirefoxBinary(log_file=f)
```

Comment:
When I pass `abs_log_file_path`, I get told that `AttributeError: 'str' object has no attribute 'fileno'`.

Looking at the [docs](https://seleniumhq.github.io/selenium/docs/api/py/webdriver_firefox/selenium.webdriver.firefox.firefox_binary.html), there appears to be no way to pass a path to a file, so I guess we're stuck with sending the file handle.